### ISSUE TITLE
dither-depth=auto is broken on vo=gpu-next.

### ISSUE THREAD

**Issue by ghost** [mpv-player/mpv#11862]
### Important Information
- mpv version: [shinchiro 20230 6240a1b00](https://github.com/shinchiro/mpv-winbuild-cmake/releases/tag/20230630)
- Platform and Version: Windows 11 22621.1928
- Source of the mpv binary: [shinchiro](https://github.com/shinchiro/mpv-winbuild-cmake/releases/tag/20230630)

### Reproduction steps
Use `vo=gpu` and `dither-depth=auto` to watch any(?) video, preferably with an 8-bit monitor or forced 8-bit output in the GPU driver. Observe the log or info page and what it says about dithering. Then contrast with what happens when you use `vo=gpu-next` and `dither-depth=auto`.

### Expected behavior
`dither-depth=auto` should always result in 8-bit dithering for an 8-bit output.

### Actual behavior
On my 8-bit monitor, 8-bit dithering fine for `vo=gpu`, regardless of the `gpu-api`  setting. But, when switching to `vo=gpu-next`, the behavior changes. With `gpu-api=d3d11`, the bit depth is incorrectly set to 10. And with `gpu-api=vulkan`, no dithering is performed at all. Both do work fine if `dither-depth=8` is specified manually.

### Log file
`vo=gpu` https://pomf2.lain.la/f/xhm706vw.txt
`vo=gpu-next -gpu-api=d3d11` https://pomf2.lain.la/f/takytfgg.txt
`vo=gpu-next -gpu-api=vulkan` https://pomf2.lain.la/f/npq8090s.txt

### Sample files
https://pomf2.lain.la/f/k1fahjqe.webm

**Comment by hooke007** [comment]
You have to manually set depth with vulkan because https://github.com/mpv-player/mpv/blob/0bfafd2451b25d25d3a8249373dc63b5d5e94379/video/out/vulkan/context.c#L275-L278

**Comment by haasn** [comment]
Working as intended. You are getting a 16-bit backbuffer with vulkan and a 10-bit backbuffer with d3d11. Dithering is not needed at 16 bit depth.

**Comment by sizumam** [comment]
That makes sense for intermediate rendering steps, but couldn't naively sending a 16-bit texture to be displayed at the end cause banding if it's not explicitly dithered to 8 bits first? That happens to me in Krita anyway.

**Comment by sizumam** [comment]
Off-topic, but will libplacebo's blue and white noise dithering ever be exposed in mpv?

**Comment by hooke007** [comment]
bluenoise →→ `dither=fruit`

**Comment by natural-harmonia-gropius** [comment]
> cause banding

`auto` results banding to me too, 8 is fine.

**Comment by Raza-0007** [comment]
> Working as intended. You are getting a 16-bit backbuffer with vulkan and a 10-bit backbuffer with d3d11. Dithering is not needed at 16 bit depth.


I have a similar issue, on my 8 bit (6 bit + FRC) display , with dither-depth=auto, vo=gpu-next dithers at 10 bit on all videos, while vo=gpu dithers at the expected 8 bit. 

Is this a bug or is this how gpu-next is supposed to dither?

Here are my tests, keeping everything else exact same, except the vo and gpu-api:

--vo=gpu-next --dither-depth=auto --gpu-api=vulkan ------> Dithering at 10 bits
--vo=gpu-next --dither-depth=auto --gpu-api=d3d11 ------> Dithering at 10 bits

--vo=gpu --dither-depth=auto --gpu-api=vulkan ------> Dithering at 8 bits
--vo=gpu --dither-depth=auto --gpu-api=d3d11 ------> Dithering at 8 bits

This is on Windows 10, mpv 0.35.0-436-g1c82d6ae (June 11 2023) bundled with the latest SMPlayer.

I can post logs if needed, but since it happens on all the videos, it should be easy to replicate on any 8 bit display.

~Raza

**Comment by hooke007** [comment]
@Raza-0007 Your question's answer is there https://github.com/mpv-player/mpv/issues/10881

**Comment by sizumam** [comment]
@haasn's [last reply there](https://github.com/mpv-player/mpv/issues/10881#issuecomment-1318964445) seems to contradict his [first reply here](https://github.com/mpv-player/mpv/issues/11862#issuecomment-1614604030). Am I dumb?

**Comment by Raza-0007** [comment]
@hooke007 Thanks. I did not consider checking as far back at Nov 2022. I thought this was a recent change. 

I am not an expert on video rendering engines, but I thought that dithering to 10 bits on a 8 bit display will cause quality issues. Something like over dithering. But reading the other discussion, @haasn seems to say that it should not cause any issues. I will of course take his word as he obviously knows more than I do about this.

@sizumam This is what I thought that when dether-depth=auto, it should either choose the exact bit depth of the display or otherwise default to 8 bit. This is why I thought it was a bug in gpu-next as it was dithering to 10 bits on auto on my 8 bit display. If it does not cause quality issues then I am ok with it.

**Comment by Raza-0007** [comment]
I did another test on an 8-bit (H264) encoded video, on my 8 bit display. vo=gpu-next is still dithering to 10 bits!

Again I am not an expert, but isn't this a bit pointless! Neither the video has nor the display is capable of displaying more than 8 bits of information, what would be gained by dithering it to 10 bits!

And yet again if someone can verify that it does not effect video quality, then I am ok with it.

~Raza

**Comment by hooke007** [comment]
> I thought this was a recent change.

As I said, what you did is what I did in that issue. It's not a recent change.

**Comment by Raza-0007** [comment]
@hooke007 What I meant was that I did a search on this issue before posting, but I did not go beyond the last 3 months, as I thought it was a recent change.

**Comment by sizumam** [comment]
> Again I am not an expert, but isn't this a bit pointless! Neither the video has nor the display is capable of displaying more than 8 bits of information, what would be gained by dithering it to 10 bits!

It would reduce rounding errors in intermediate steps. But that should just be a matter of it using a 10-bit FBO, and dithering should still be to 8 bits at the end I would think...

**Comment by haasn** [comment]
> That makes sense for intermediate rendering steps, but couldn't naively sending a 16-bit texture to be displayed at the end cause banding if it's not explicitly dithered to 8 bits first? That happens to me in Krita anyway.

Do you see visual banding? Test on a 16-bit gradient, e.g.:

![Gradient-16bit](https://github.com/mpv-player/mpv/assets/1149047/e879b332-5850-4c68-9897-92b481ff9bf3)

https://github.com/mpv-player/mpv/assets/1149047/f424d0fb-dcdb-4b10-aada-f71fa8d592ac

If you do, that means it's a driver/platform bug. The system should not advertise a texture format higher than the monitor depth if it does not perform internal dithering..

**Comment by natural-harmonia-gropius** [comment]
I don't know is screenshot can show banding.

win11 d3d11 nvidia 536.40

depth 8
![image](https://github.com/mpv-player/mpv/assets/50797982/7907d5b3-be1a-475e-9dfc-d41ef86171b6)
depth auto (10bit)
![image](https://github.com/mpv-player/mpv/assets/50797982/57dca958-c6b8-42d9-9f26-e9afa6cc9fe4)

**Comment by haasn** [comment]
Screenshot is irrelevant, unless you can do a capture of the video signal coming into the monitor (e.g. HDMI capture card). Do you see banding visually?

**Comment by natural-harmonia-gropius** [comment]
> Do you see banding visually?

yes, very noticeable in dark aera.

**Comment by haasn** [comment]
And does the above patch fix it?

**Comment by Raza-0007** [comment]
> Of course, this would still pick 10-bit backbuffers if available. There's almost fundamentally no way around this, because we don't know the true monitor bit depth and have to rely on the compositor not lying to us.

@haasn Since you are here, I just wanted to ask, as my original issue was not related to 16-bit backbuffer. Below are two log files of a 8 bit video that I am playing on a 8 bit display. All options are the same except the --vo and --dither-depth=auto for both.

[vo=gpu-next.txt](https://github.com/mpv-player/mpv/files/11940281/vo.gpu-next.txt)

[vo=gpu.txt](https://github.com/mpv-player/mpv/files/11940278/vo.gpu.txt)

--gpu-next chooses to dither to 10 bits, while --gpu dithers to 8 bits.

From what I have read, you are saying that --gpu-next does this by design and there is no quality loss in the output by dithering to 10 bits for this setup. Is this correct?

**Comment by haasn** [comment]
> From what I have read, you are saying that --gpu-next does this by design and there is no quality loss in the output by dithering to 10 bits for this setup. Is this correct?

Assuming your graphics driver is working correctly, yes.

**Comment by sizumam** [comment]
> And does the above patch fix it?

@haasn no. All 12 results are the same on my machine with that patch.
`[   0.255][v][vo/gpu-next/libplacebo] Picked surface configuration 22: VK_FORMAT_R16G16B16A16_UNORM + VK_COLOR_SPACE_SRGB_NONLINEAR_KHR`
https://pomf2.lain.la/f/p9zb7tm.txt

**Comment by hooke007** [comment]
> Screenshot is irrelevant, unless you can do a capture of the video signal coming into the monitor (e.g. HDMI capture card). 

His screenshots showed clear banding in dark area...

**Comment by Raza-0007** [comment]
> Assuming your graphics driver / compositor is working correctly, yes.

Thank You!

**Comment by haasn** [comment]
> > And does the above patch fix it?
> 
> @haasn no. All 12 results are the same on my machine with that patch. `[ 0.255][v][vo/gpu-next/libplacebo] Picked surface configuration 22: VK_FORMAT_R16G16B16A16_UNORM + VK_COLOR_SPACE_SRGB_NONLINEAR_KHR` https://pomf2.lain.la/f/p9zb7tm.txt

Are you sure you had the patch applied? It should be impossible to pick that format combination with the above patch.

**Comment by llyyr** [comment]
https://github.com/mpv-player/mpv/blob/master/video/out/vulkan/context.c#L277

`--dither-depth=auto` is not implemented for the vulkan context, you should use d3d11 if you want it to be auto detected, or set it to 8 manually.

**Comment by hooke007** [comment]
> --dither-depth=auto is not implemented for the vulkan context, you should use d3d11 if you want it to be auto detected, or set it to 8 manually.

It was answered here https://github.com/mpv-player/mpv/issues/11862#issuecomment-1614454558

**Comment by haasn** [comment]
> https://github.com/mpv-player/mpv/blob/master/video/out/vulkan/context.c#L277
> 
> `--dither-depth=auto` is not implemented for the vulkan context, you should use d3d11 if you want it to be auto detected, or set it to 8 manually.

This is irrelevant, it affects only `--vo=gpu`, but OP was talking about `--vo=gpu-next`.

**Comment by Raza-0007** [comment]
I believe that whoever is in charge of the mpv manual needs to update this section on `--dither-depth` here:

https://mpv.io/manual/stable/#options-dither-depth

Since the `--dither-depth` behavior is different for each supported `--vo` and `--gpu-api`, a few lines of explanation in the manual will put people's mind at ease, and they will know that what they are seeing in the log is by design, and not a bug in mpv. 

For example, it is generally known that if dithering is done at any bit depth other than the connected display's default bit depth, it will lead to ugly output. But if this is not true, and there is no quality loss in dithering to 10 bits with `--vo=gpu-next`, then this should be specified in the manual.

Also the manual currently says that on `--dither-depth=auto` if bit depth cannot be detected, then 8 bit per component are assumed, which obviously is not true for `--vo=gpu-next`, as it always defaults to 10 bits, so this section needs to be updated.

~Raza

**Comment by haasn** [comment]
> Also the manual currently says that on `--dither-depth=auto` if bit depth cannot be detected, then 8 bit per component are assumed, which obviously is not true for `--vo=gpu-next`, as it always defaults to 10 bits, so this section needs to be updated.

This is not true, though. `vo_gpu_next` only dithers to 10 bits when the surface is **detected** as 10-bit. The bit about the bit depth not being detectable only applies to OpenGL. On Vulkan, we *always* know the exact bit depth of the backbuffer.

The manual description is entirely consistent with the observed behavior.

**Comment by Raza-0007** [comment]
> 
> This is not true, though. `vo_gpu_next` only dithers to 10 bits when the surface is **detected** as 10-bit.  
On Vulkan, we _always_ know the exact bit depth of the backbuffer.

But then why is `--vo=gpu-next` with `--gpu-api=vulkan` detecting both my internal and external displays as having a bit depth of 10 bits, when both are only 8 bit capable displays? This is the part that has been bothering me since I discovered it in the log entry.

**Comment by haasn** [comment]
> > This is not true, though. `vo_gpu_next` only dithers to 10 bits when the surface is **detected** as 10-bit.
> > On Vulkan, we _always_ know the exact bit depth of the backbuffer.
> 
> But then why is `--vo=gpu-next` with `--gpu-api=vulkan` detecting both my internal and external displays as having a bit depth of 10 bits, when both are only 8 bit capable displays? This is the part that has been bothering me since I discovered it in the log entry.

Because your _compositor_ is advertising us a 10-bit backbuffer. Most compositors will do internal dithering down to 8 bit before sending it out for scanout.

**Comment by Raza-0007** [comment]
@haasn Thanks for the explanation. Since you have already confirmed that all this up dithering and down dithering do not effect video quality, then I am OK with it. 

And thanks for your wonderful work on gpu-next and libplacebo.

~Raza

**Comment by Touceire** [comment]
>do not effect video quality

But it absolutely does.
`mpv --no-config --vo=gpu-next --gpu-api=vulkan --dither-depth=auto` causes terrible banding while `mpv --no-config --vo=gpu-next --gpu-api=vulkan --dither-depth=8` is fine.

**Comment by Raza-0007** [comment]
@Touceire I have not noticed any banding issues, but then I have `--deband=yes` enabled among the options I pass to mpv, so that might be smoothing out the banding for me. I will do some tests with `--deband` disabled to see if I notice banding issues.

**Comment by llyyr** [comment]
> This is irrelevant, it affects only --vo=gpu, but OP was talking about --vo=gpu-next.

Right, I incorrectly assumed that this was broken because of that TODO but I can also reproduce this on Linux. When `dither-depth=auto` is set with `gpu-api=vulkan`, no dithering happens at all. But if I set `gpu-api=opengl` I do see logs about dithering and in libplacebo output:

`[   0.170][d][vo/gpu-next/libplacebo] [ 86] // pl_shader_dither`

Here's both logs
[dither_auto_vulkan.log](https://github.com/mpv-player/mpv/files/11953344/dither_auto_vulkan.log)
[dither_auto_opengl.log](https://github.com/mpv-player/mpv/files/11953346/dither_auto_opengl.log)

Another thing to add is that plplay does do 8 bit dithering by default on the same file, and even picks the right surface configuration while mpv does not

plplay:

```
Available surface configurations:
    0: VK_FORMAT_B8G8R8A8_SRGB                  VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
    1: VK_FORMAT_B8G8R8A8_UNORM                 VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
Picked surface configuration 1: VK_FORMAT_B8G8R8A8_UNORM + VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
Codec: hevc (HEVC (High Efficiency Video Coding))
Using hardware frame format: vaapi
Spent 38.476 ms translating SPIR-V
Spent 10.853 ms generating shader LUT
Dithering to 8 bit depth
```

mpv:

`[   0.080][v][vo/gpu-next/libplacebo] Picked surface configuration 13: VK_FORMAT_R16G16B16A16_UNORM + VK_COLOR_SPACE_SRGB_NONLINEAR_KHR`

**Comment by haasn** [comment]
@llyyr And with the above patch to disable R16G16B16A16?

**Comment by haasn** [comment]
> @llyyr And with the above [patch to disable R16G16B16A16](https://code.videolan.org/videolan/libplacebo/-/merge_requests/494)?

> Another thing to add is that plplay does do 8 bit dithering by default on the same file, and even picks the right surface configuration while mpv does not

Impossible to verify without log but there's a good chance this is because of it using X instead of Wayland (depending on how your GLFW is configured), and Xwayland not supporting high-bit-depth backbuffers.

**Comment by llyyr** [comment]
With the patch, it dithers to 10 bit output.

`[   0.183][v][vo/gpu-next/libplacebo] Dithering to 10 bit depth`

> there's a good chance this is because of it using X instead of Wayland (depending on how your GLFW is configured), and Xwayland not supporting high-bit-depth backbuffers.

This is correct, I was able to verify it by running plplay with SDL2 backend and making sure it was running under Wayland natively. `SDL_VIDEODRIVER=wayland plplay -a sdl2-vk -v [video] > plplay.log`

`Dithering to 10 bit depth` [plplay.log](https://github.com/mpv-player/mpv/files/11956472/plplay.log)

```
Available surface configurations:
    0: VK_FORMAT_B8G8R8A8_SRGB                  VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
    1: VK_FORMAT_B8G8R8A8_UNORM                 VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
    2: VK_FORMAT_R8G8B8A8_SRGB                  VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
    3: VK_FORMAT_R8G8B8A8_UNORM                 VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
    4: VK_FORMAT_R4G4B4A4_UNORM_PACK16          VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
    5: VK_FORMAT_B4G4R4A4_UNORM_PACK16          VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
    6: VK_FORMAT_R5G6B5_UNORM_PACK16            VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
    7: VK_FORMAT_B5G6R5_UNORM_PACK16            VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
    8: VK_FORMAT_R5G5B5A1_UNORM_PACK16          VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
    9: VK_FORMAT_B5G5R5A1_UNORM_PACK16          VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
    10: VK_FORMAT_A1R5G5B5_UNORM_PACK16          VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
    11: VK_FORMAT_A2R10G10B10_UNORM_PACK32       VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
    12: VK_FORMAT_A2B10G10R10_UNORM_PACK32       VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
    13: VK_FORMAT_R16G16B16A16_UNORM             VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
    14: VK_FORMAT_R16G16B16A16_SFLOAT            VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
Picked surface configuration 11: VK_FORMAT_A2R10G10B10_UNORM_PACK32 + VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
Codec: h264 (H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10)
Using hardware frame format: vaapi
Spent 38.619 ms translating SPIR-V
Spent 11.706 ms generating shader LUT
Dithering to 10 bit depth
```

**Comment by haasn** [comment]
> With the patch, it dithers to 10 bit output.

And that gets rid of visible banding? (Does the driver dither from 10 to 8 bits internally?)

**Comment by llyyr** [comment]
I see more banding when mpv is doing 10 bit dithering than when it is doing 8 bit dither, so I don't think the driver is doing any dithering internally.

**Comment by haasn** [comment]
In that case, i.e. assuming the driver isn't dithering internally, then In theory, mpv doing 10 bit dithering vs mpv doing 16 bit dithering (or even doing no dithering at all) shouldn't make any difference.

Sounds like this may be a severe bug in whatever Wayland compositor you're using. Which one is it?

**Comment by llyyr** [comment]
My compositor does internally dither down to 10 to 8 bits internally when I set `render_bit_depth 10` (ctrl+f 'render_bit_depth' in case it's relevant https://man.archlinux.org/man/sway-output.5.en)

mpv doing 8 bit dithering vs 10 bit dithering has no difference in the output to my screen when my render_bit_depth is set to 10.

In my previous test, it was set to 8.

**Comment by haasn** [comment]
FWIW I can't reproduce this on sway with amdgpu/mesa. I get offered only 10-bit formats, no 16-bit, and don't get visible banding at all, even when completely disabling dithering (on the 10-bit sample).

So on my end, the wayland compositor / graphics driver work together to correctly dither output. Tested on sway, weston, plasma.

**Comment by haasn** [comment]
I tested sway with `render_bit_depth 8` (though this was presumably also the deafult in my previous test) and I *still* get dithering when picking a 10-bit backbuffer on an 8-bit screen. The *only* configuration that gives me banding is combining an 8-bit backbuffer with no dithering.

**Comment by haasn** [comment]
After a substantial amount of debugging with @llyyr on IRC I came to the conclusion that everything works correctly on my end - wlroots gets a 16-bit buffer from vulkan, then forwards this for scanout to the display. Even if the display is artificially forced into 8-bit mode, I see no dithering, confirming that the amdgpu kernel driver is performing dithering as expected (and as confirmed in the amdgpu DC source code).

So, in conclusion, I have no idea why it doesn't work for some people. Some bug somewhere, but only god knows where.

**Comment by dokur0chan** [comment]
Sorry for the necro, but I have a curiosity question regarding driver dithering after reading this thread.

Even if driver dithering is working as intended, is there any reason why the end-user shouldn't set `--dither-depth` to their display's bit-depth regardless? Is there any reason to believe that the compositor's built-in dithering is "superior" to whatever dither shaders mpv offers? (or vice-versa?)

On `d3d11`, `--dither-depth=auto` is already "layering" a 10-bit `fruit` dither pattern alongside driver dithering, so there shouldn't be any visual oddities with using `--dither-depth=8` on an 8 bpc display, no? At the very least, there technically should be no perceptible differences even on synthetic test content. But if compositor dithering is indeed superior to mpv's available dither shaders, then from what I'm understanding, dithering should be kept at `--dither-depth=no` instead of `auto` or `<insert display bpc here>`.

**Comment by llyyr** [comment]
> Is there any reason to believe that the compositor's built-in dithering is "superior" to whatever dither shaders mpv offers?

No, driver dithering is probably always worse than what mpv can offer. But it's extremely difficult to tell the difference. Though at this point, after testing on both Intel/AMD gpus, across Windows/X11/Wayland, I've considered that maybe my driver *is* dithering and it's so bad at it that I think it's not doing anything at all.

**Comment by kasper93** [comment]
> I've considered that maybe my driver is dithering and it's so bad at it that I think it's not doing anything at all.

I think dithering is pretty simple process to break it that bad. But who knows, maybe they implement dither=clip :P

Or maybe it is by design not dithering in some cases to allow users (mpv) dither instead. Double dithering might not be good idea.

**Comment by dokur0chan** [comment]
> Or maybe it is by design not dithering in some cases to allow users (mpv) dither instead. 

Well, that would be awfully considerate of AMD if that's true...

> Double dithering might not be good idea.

I'm assuming that you're already "double dithering" when using `d3d11` and setting a `--dither-depth` in mpv, since as concluded in this thread, `d3d11` doesn't have banding regardless of `--dither-depth` settings (I can reproduce this on my own setup as well). But intuitively, I feel like there would be no issues with stacking dither patterns if both of them are static. If you have a FRC monitor, your display is already applying temporal dithering to achieve a higher color depth, so having two static dither patterns "under" that shouldn't cause any issues... I think. 

That's part of the reason why I asked anyways. There's too many hypotheticals and assumptions when discussing this topic, I'm almost content with just saying "if you don't notice anything wrong, don't worry about it".

**Comment by haasn** [comment]
> Sorry for the necro, but I have a curiosity question regarding driver dithering after reading this thread.
> 
> Even if driver dithering is working as intended, is there any reason why the end-user shouldn't set `--dither-depth` to their display's bit-depth regardless? Is there any reason to believe that the compositor's built-in dithering is "superior" to whatever dither shaders mpv offers? (or vice-versa?)
> 
> On `d3d11`, `--dither-depth=auto` is already "layering" a 10-bit `fruit` dither pattern alongside driver dithering, so there shouldn't be any visual oddities with using `--dither-depth=8` on an 8 bpc display, no? At the very least, there technically should be no perceptible differences even on synthetic test content. But if compositor dithering is indeed superior to mpv's available dither shaders, then from what I'm understanding, dithering should be kept at `--dither-depth=no` instead of `auto` or `<insert display bpc here>`.

If you/we know true bit depth of the monitor connection I would always dither to that depth, just to be on the safe side. Double-dithering should never be an issue - because the values are always quantized, the error will be 0 for the next dithering pass.

**Comment by kasper93** [comment]
Maybe we should update manual, to not scary users about
> Note that the depth of the connected video display device cannot be detected. Often, LCD panels will do dithering on their own, which conflicts with this option and leads to ugly output.

**Comment by haasn** [comment]
Yeah, feel free to remove it.

**Comment by llyyr** [comment]
So after some comprehensive testing on this issue, I've reached the conclusion that it's somehow a "Vulkan" bug.

I tested on both Intel and AMD, on both Linux and Windows. (Assume all tests are with --dither-depth=auto, unless stated otherwise)

With Vulkan, I always see banding unless mpv is forced into picking an 8 bit backbuffer.

With opengl on Linux, there's no banding even when it picks a 10 bit backbuffer.

With d3d11 on Windows, there's no banding even when it picks a 10 bit backbuffer.

I've managed to eliminate mesa/gpu drivers entirely, because this issue is present across RDNA2, and multiple generations of Intel iGPU, and going by reports in this issue, on nvidia too. It's _possible_ that multiple gpu drivers have the same bug, but highly unlikely.

The only common theme here is the use of Vulkan API, so the issue should be somehow reported to them. However, I don't know what the appropriate place for "Vulkan" bugs is, since from my understanding, the implementation for Vulkan should be different on both Windows and Linux, but the same bug is present.

If anyone has any idea what the right place to report this Vulkan issue is, please let me know.

**Comment by kasper93** [comment]
So what do we do here? Clearly the expectation of mpv that 10/16 backbuffers will be dithered down to display bitness is not true.

I see three possible solutions

A) Read EDID of the current display and make `--dither-depth=auto` smarter about what it should dither to

B) Assume that we don't have this information and it is on user to correctly configure their playack.

C) Enable dithering in driver, so the backbuffers are correctly dithered when sending to display
For that there are API, see
https://gpuopen-librariesandsdks.github.io/adl/group__DISPLAYAPI.html#ga057523cefb56aaabee51b67a9bd2a6d3
https://gpuopen-librariesandsdks.github.io/adl/group__define__dither__states.html
Also it is not unexpected that dithering is disabled by default.
> ADL_DL_DISPLAY_DITHER_DRIVER_DEFAULT
> Use default driver settings for dithering. Note that the default setting could be dithering disabled.

nvidia also has an API for controlling dithering, although they don't provide public documentation on that, there are API tables with functions like that that basically do the same thing as ADL one.
```
NvAPI_GPU_GetDitherControl 0x932AC8FB
NvAPI_GPU_SetDitherControl 0xDF0DFCDD
```

Not sure about Intel.

**Comment by lextra2** [comment]
EDID can also be wrong. LG OLEDs allow for a bit depth of 12 (under Windows GPU control panel settings) which then is internally dithered by the TV to 10bit.

**Comment by kasper93** [comment]
At this point you have to configure it yourself, how can mpv know that?

**Comment by lextra2** [comment]
Nvidia drivers also [don't dither by default.](https://github.com/ledoge/novideo_srgb?tab=readme-ov-file#dithering) if Full-RGB output is used.

+1 for defaulting MPV to `dither-depth=no`

**Comment by llyyr** [comment]
> At this point you have to configure it yourself, how can mpv know that?

I think disabling 10/16 bit backbuffers for SDR rendering makes sense, so hooking up that libplacebo API with mpv should deal with some of this. Otherwise I think it's upto the user to configure dither-depth.

**Comment by lextra2** [comment]
With Windows 11, the compositor can dither, too, if Auto Color Management (ACM) is used.
That would be in total 4 steps of dithering (MPV > GPU > Windows > TV/Monitor

Pure madness

**Comment by kasper93** [comment]
I've added support for `--dither-depth=auto` for gpu-next, on supported APIs. https://github.com/mpv-player/mpv/pull/13646

**Comment by Raza-0007** [comment]
I just wanted add this here in case somebody else has noticed this issue. Since the change where `dither-depth=auto` now defaults to 8 on `gpu-api=vulkan` with `vo=gpu-next`, the colors have become a touch less vibrant, and the image a touch less sharp on my system. For example, reds are slightly less redder etc. and image looks less vibrant.

I had earlier asked on this same thread whether dithering to 10 bits on my 8 bit displays will compromise image quality. And I was told by haasn that as long as my drivers and compositor was working correctly, it will not. So I have been using `dither-depth=auto` (10 bits on my system) since then without any issues. 

But now with the change I have noticed the slight loss of image quality when mpv dithers to 8 bits. Changing `dither-depth=10` fixes the issue, and restores the colors and image quality back. The tests were done with `gpu-api=vulkan` on `vo=gpu-next` and keeping everything else constant, just changing the `dither-depth` between auto and 10.

I do not have any explanation other than this that mpv always picks a 10 bit surface configuration on my system with 8 bit displays, so maybe mpv needs to dither to 10 bits to give me the best image quality.

By the way, I never had any banding issues that others here have experienced, so if someone was experiencing banding, then a slight loss of image quality is acceptable. 

~Raza

**Comment by wiwaz** [comment]
Placebo.

**Comment by lextra2** [comment]
@Raza-0007 
What OS and what GPU?
If on windows, go to System32 folder and doubleclick on vulkaninfo.exe and post the output here

**Comment by haasn** [comment]
> I just wanted add this here in case somebody else has noticed this issue. Since the change where `dither-depth=auto` now defaults to 8 on `gpu-api=vulkan` with `vo=gpu-next`, the colors have become a touch less vibrant, and the image a touch less sharp on my system. For example, reds are slightly less redder etc. and image looks less vibrant.
> 
> I had earlier asked on this same thread whether dithering to 10 bits on my 8 bit displays will compromise image quality. And I was told by haasn that as long as my drivers and compositor was working correctly, it will not. So I have been using `dither-depth=auto` (10 bits on my system) since then without any issues.
> 
> But now with the change I have noticed the slight loss of image quality when mpv dithers to 8 bits. Changing `dither-depth=10` fixes the issue, and restores the colors and image quality back. The tests were done with `gpu-api=vulkan` on `vo=gpu-next` and keeping everything else constant, just changing the `dither-depth` between auto and 10.
> 
> I do not have any explanation other than this that mpv always picks a 10 bit surface configuration on my system with 8 bit displays, so maybe mpv needs to dither to 10 bits to give me the best image quality.
> 
> By the way, I never had any banding issues that others here have experienced, so if someone was experiencing banding, then a slight loss of image quality is acceptable.
> 
> ~Raza

Screenshots?

**Comment by Raza-0007** [comment]
@lextra2, This is on Windows 11, with Nvidia 1070 GTX

Here is the vulkaninfo
[vulkaninfo.txt](https://github.com/mpv-player/mpv/files/14733806/vulkaninfo.txt)

@haasn, I cannot notice the issue in screenshots. Could be due to the jpg compression. Since screenshots are unscaled, could the scalers in my config have anything to do with this? I have a simple config, nothing special, but everything was enabled during the test, I only changed the `dither-depth`. 
[mpv config.txt](https://github.com/mpv-player/mpv/files/14733807/mpv.config.txt)

But if none of you are experiencing this, then it might just be me. Changing `dither-depth=10` fixes it for me for now. I will continue doing more tests on it to rule out bias caused by my eyes. 

Thanks.

~Raza

**Comment by haasn** [comment]
> @haasn, I do not notice the issue in screenshots. Could be due to the jpg compression. Since screenshots are unscaled, could the scalers in my config have anything to do with this? I have a simple config, nothing special, but everything was enabled during the test, I only changed the dither-depth.

Make a screenshot of your screen (printscreen), not a screenshot inside mpv.

**Comment by lextra2** [comment]
`format = FORMAT_A2B10G10R10_UNORM_PACK32`
`colorSpace = COLOR_SPACE_SRGB_NONLINEAR_KHR`
Your GPU driver supports vulkan with 10bit, so if you've selected 10bit in the gpu control panel, you should use `dither-depth=10` which you manually have to set yourself since Vulkan has no extension to check for the current displays bit depth.

**Comment by Raza-0007** [comment]
@haasn, I have taken multiple screenshots (printscreen) with different types of media for the last hour, but strangely, I cannot see the difference in the screenshots. I can notice it when the media is played. It is only a slight difference, image with 10 bit dithering is slightly more smoother... better. But I guess, since screenshots don't show this even after zooming into them, so it must be just my eyes, and not mpv. Here are four samples in the zip archive, but they are identical.

[Screenshots.zip](https://github.com/mpv-player/mpv/files/14734049/Screenshots.zip)


@lextra2,  This 10 bit format `FORMAT_A2B10G10R10_UNORM_PACK32` is exactly what mpv selects for me every time. And previously at `dither-depth=auto` it used to dither at 10 bit too. Now with 8 bit dithering, the image is not bad, I think my eyes are used to the 10 bit dithering, so I can tell the difference. It only a very slight difference. But I have set my dither depth at 10 for now, which makes everything the same as it used to be.

But thank you, both of you for trying to help. It is clearly an issue on my system only, and most likely it is just me. I thought I should mention it here in case someone else was experiencing something similar, but I guess no one else is :blush:

~Raza

**Comment by lextra2** [comment]
>I have taken multiple screenshots (printscreen) with different types of media for the last hour, but strangely, I cannot see the difference in the screenshots.

Maybe related to Nvidias dithering. You can't capture that via screenshots.
Try disabling dithering with one of these:

https://forums.blurbusters.com/viewtopic.php?f=27&t=12953

https://forums.guru3d.com/threads/nvidia-and-dithering-controls-how-to-enable.436621/#post-5885004

Note: Any of those registry changes requires a reboot to take effect.

**Comment by Raza-0007** [comment]
@lextra2, thanks for the links, some good info there. I see that the registry fixes mentioned on that thread are for fixing eye strain caused by temporal dithering, therefore, I am hesitant to try them or the colorcontrol utility on my system, because my issue is trivial, and can be easily fixed by forcing `dither-depth=10` in config. Everything else is working fine. I don't want to inadvertently introduce a bigger issue by editing these settings. 

~Raza
